### ISSUE TITLE
On-Policy Distillation Feature Proposal

### ISSUE THREAD

**Issue by JacobHelwig** [PrimeIntellect-ai/prime-rl#2189]
I've been looking through the OPD code and wanted to suggest a few new features: VLM teacher, Multi-teacher OPD, top-k distillation loss. I'd be happy to work on any of these if it would be useful.

# VLM Teacher

Currently, VLM distillation is limited to SFT. To extend to OPD, a custom endpoint supporting TITO and multi-modal prompts is needed for the teacher server. One way could be by extending `OpenAIServingChatWithTokens.create_chat_completion_with_tokens` to handle multi-modal prompts:

1. Endpoint requests are a mix of the multi-modal prompt messages (raw text + images) and response tokens generated by the student.
2. Multi-modal prompt messages are preprocessed to prompt token IDs and `multi_modal_data` using the `OpenAIServingChat.render_chat_request` method.
3. Student response token IDs are concatenated to the prompt token IDs in the engine prompt, `multi_modal_data` is also passed in.

# Top-k teacher logits

Currently, only Thinking labs-style OPD is supported, which uses policy gradient updates with the negative of k1 KL estimator as the reward. This gives training signal on how the logprob of the sampled token should be adjusted, but does not provide any information as to how other logprobs should be adjusted.

GKD-style OPD (https://arxiv.org/abs/2306.13649) instead uses exact KL over the full logits as a supervised loss (i.e., no PG, directly backprop the loss). Since transport of a `[B,S,V]` tensor between inference/train engines has high overhead, other RL frameworks instead use top-k distillation loss:

1. NeMo RL: https://github.com/NVIDIA-NeMo/RL/blob/d430b3c675c954f6b03809465707ab85c2e8a94f/examples/configs/distillation_math.yaml#L13
2. verl: https://github.com/verl-project/verl/blob/26799888f9235dd25cfabd00a53ff841059ad61b/verl/trainer/config/distillation/distillation.yaml#L29-L30

Main change set would be:

1. Post-processing of prompt logprobs in `compute_teacher_logprobs` and add post-processing of top-k token IDs.
2. Transport between orchestrator and trainer to handle top-k logprobs and IDs.
3. Gather student logprobs for each micro batch in `train.py` at top-k teacher IDs.
4. Add a top-k distillation `loss_fn` in `loss.py` 

# Multi-teacher OPD

When training a student in multiple environments, knowledge can be distilled from domain-specific teachers. For example, for coding problems, distillation logprobs come from a coding teacher, and for math problems, they come from a math teacher. This has been used by:

1. MiMo-V2-Flash (Sec 4.4 in https://arxiv.org/abs/2601.02780)
2. GLM-5 (Sec 3.5 in https://arxiv.org/abs/2602.15763)
3. Nemotron-Cascade 2 (Sec 4.4 in https://arxiv.org/abs/2603.19220)

Main change set would be:

1. Multiple `teacher_inference` configs and `TeacherModelConfig` configs in `RLConfig`
2. Start one client per teacher in `rl.py`
3. Add one inference pool per teacher in `orchestrator.py`
4. Route `sample` to `client` based on task for that sample.

**Comment by mikasenghaas** [comment]
hey @JacobHelwig! super nice ideas.

1. i think we should support this, yes (cc @hallerite to confirm). in fact, in general we should avoid to support features only on a few code branches (e.g. only rl, but not opd). i think we might need some refactors around the current teacher configs 
2. this seems cool, but not sure if super high prio for us atm (cc @faresobeid)
3. i like this a lot, i think from a framework point of view a general `list[InferenceConfig]` might be super useful (been wanting this for local llm judges for a while but didn't get around to this). do you have a design proposal here? i think we could aim to deprecate `teacher_inference` in favor of general inference config list + explicit config to switch into opd (maybe even entrypoint?!) + a way to identify clients with base urls

**Comment by JacobHelwig** [comment]
Thanks, @mikasenghaas! I have a plan for point 1 and point 3 below, happy to work on either one if it would be helpful.

## Multi-teacher OPD

Here is a design proposal for MOPD. I used a `dict[str,InferenceConfig]`, since it makes mapping teachers to environments a bit easier. To map clients to `base_url`, we create `TeacherModelConfig` configs for each `InferenceConfig`, and add them to a `dict[str, TeacherModelConfig]` alongside any external teacher servers: 

1. `orchestrator.teacher_models: dict[str, TeacherModelConfig]`: Keys are `teacher_id`s. Populated explicitly by the user for external teachers, or auto-populated for ones spawned in `rl.py` (see point 2). Replaces `orchestrator.teacher_model`.
2. `teacher_inference: dict[str, InferenceConfig]`: Keys are `teacher_id`s. A validator adds a `orchestrator.teacher_models[teacher_id]` entry for each. `rl.py` then spawns each of the inference servers.
3. `orchestrator.opd.teachers: dict[str, str]`: Maps `env_name` --> `teacher_id`.
4. Teacher inference pool: In `orchestrator.py`, call `setup_inference_pool` for each `teacher_models` entry, then resolve `opd.teachers` into `dict[env_name, InferencePool]`.
5. `compute_teacher_logprobs`: extend to route `TrainingSample`s to teacher inference pools by `env_name`

### Example Config

```toml
# Student rollout server
[inference]
[inference.model]
name = "Qwen/Qwen2.5-3B-Instruct"

[inference.parallel]
dp = 4

# Local teacher: rl.py spawns it
[teacher_inference.code.parallel]
dp = 2
[teacher_inference.code.model]
name = "Qwen/Qwen2.5-Coder-7B-Instruct"

# External teacher: already running elsewhere
[orchestrator.teacher_models.math.client]
base_url = ["http://localhost:8003/v1"]
skip_model_check = true
[orchestrator.teacher_models.math.model]
name = "Qwen/Qwen2.5-Math-7B-Instruct"

[orchestrator.opd.teachers]
code-env = "code"
mini-swe-agent-plus = "code"
math-env = "math"

[[orchestrator.env]]
id = "code-env"

[[orchestrator.env]]
id = "mini-swe-agent-plus"

[[orchestrator.env]]
id = "math-env"

[trainer.loss]
teacher_tau = 0.5
adv_tau     = 1.0
```


## VLM OPD

I think that the main changes would be to include multi-modal data in the inputs to the teacher server. Specifically, we'd have to retain raw images instead of deleting them in `build_vlm_image_cache` so that they could be processed using the teacher processor in an updated version of the `/chat/completions/tokens` endpoint. Although we keep the processor outputs from the student, there is no path in vLLM for passing in processor outputs. 

This also gives a cleaner path if we want to support student and teacher with different processors, although I checked that Qwen3-VL-4B-Instruct/Qwen3-VL-8B-Instruct and gemma-4-26B-A4B-it/gemma-4-31B-it use the same processors, so it seems this might be a safe assumption. The only remaining step for supporting different processors would be to handle different image tokenizations, since this could result in different number of placeholder image tokens in the prompt + completion IDs for processors with different patch/merge sizes.

I did make a PR (#2245) that attempted to add VLM OPD support following the mix of MITO and TITO mentioned in the original issue above. I've since realized that it is possible for `TrainingSample.completion_ids` to contain images in multi-turn settings, in which case the proposed solution fails.
